### ISSUE TITLE
Wifi won't connect even 'disconnect' button has shown up

### ISSUE THREAD

**Issue by Hzj-jie** [koreader/koreader#2183]
- KOReader version: Up to 06bd4d7
- Device: Kobo Aura HD
#### Issue

Two issues found in Network manager feature,
1. Device cannot connect to Wifi without password. Error message 'Networks without WPA/WPA2 encryption are not supported.'
2. After enabling Wifi connection, all SSIDs are not clickable, while one of them shows a disconnect button. But indeed no network access in this status, I need to manually disconnect and connect again.

And one more FR, it would be good if we can restore last status after resuming from device suspending.

**Comment by houqp** [comment]
> Two issues found in Network manager feature,
> 1. Device cannot connect to Wifi without password. Error message 'Networks without WPA/WPA2 
encryption are not supported.'

Yup, I was thinking most of people enable WPA encryption at home so did not bother to implement WEP and open network support. Are you using this in a public wifi environment?

> 2. After enabling Wifi connection, all SSIDs are not clickable, while one of them shows a disconnect button. But indeed no network access in this status, I need to manually disconnect and connect again.

Are you saying you still don't have internet access after the first connect attempt? If so, it sounds like a bug. I haven't run into this before. Did you see the "Connected." info message? I am planning to add a network status page from the menu to show basic info like obtained IP address.

> And one more FR, it would be good if we can restore last status after resuming from device suspending.

Yeah, this is not implemented yet, but something really nice to have.

**Comment by Hzj-jie** [comment]
1. Oh, I thought it should be even simpler. Yes, I may use a public wifi sometime.
2. No, it seems I always need to disconnect then connect. The produce step is,
- turn on wifi, and connect as normal.
- suspend by hardware button or auto suspend feature.
- resume the device, the wifi should be shutdown.
- turn on wifi, a list of SSIDs shows up. I have not seen a 'Connected' info message.
- a disconnect button shows up after one of the SSIDs.
- but in this state, no network access. I need to manually disconnect and reconnect.

**Comment by houqp** [comment]
> 1. Oh, I thought it should be even simpler. Yes, I may use a public wifi sometime.

It's simpler, but requires writing extra code :P 

> No, it seems I always need to disconnect then connect.

If it only happens after coming back from a resume, then it's expected, or rather, a known issue. The root cause is after a suspend/resume, wpa_suppliant is still connected to the AP. That's why you are seeing the disconnect button. But the IP address for the network interface is lost, that's why you can't talk to the internet. We just need to call dhcp again after the resume to reassign the IP address.

**Comment by Biep** [comment]
This may or may not be the same.  On my Kobo Aura one, with KSM, in KOReader I try to connect to a Wifi.
(1) There are lots of Wifis available, more than fit in the list shown, and the Wifi I want isn't show.  I see no option to scroll the list.
(Thinking maybe the naming is off, I try connecting to a few of the ones I see, but of course the password is not correct.  Passwordless Wifis don't work as per previous messages.)
(2) I now want to switch Wifi off again, and go to Settings -> Network, where I see "Wi-Fi connection" checked.  When I try to uncheck, it ask whether I want to switch Wifi **ON**.  I finally leave KOReader, and am asked "Switch off Wifi?"whether I want to switch Wifi off - which I do.  But going back to KOReader, the box is still checked, and I still get the same question when trying to uncheck it.  And leaving again it asks me whether I want to switch it off.  It seems Wifi is always on now, though not connected.

**Comment by Frenzie** [comment]
> (1) There are lots of Wifis available, more than fit in the list shown, and the Wifi I want isn't show. I see no option to scroll the list. (Bug 1)

Swipe left/right.

> (4) I power my ereader off and on again, but upon getting back into KOReader, the "Wi-fi connection" is still checked, and leaving it I get asked again whether I want to switch Wifi off (which, BTW, I do every time).
So, is Wifi on all the time, and draining my battery?

Sounds like it, but it wouldn't be KOReader turning it on. What does "getting back into KOReader" entail?

Also pinging @NiLuJe who owns an Aura One iirc?

**Comment by Biep** [comment]
Swipe left-right!  Yes, that works.
I had tried up-down, with one and two fingers, as the list is vertical.  Maybe allow (also) vertical swiping?

Getting back into KOReader = choosing it from KSM.

**Comment by NiLuJe** [comment]
Try the latest nightly it, it should attempt to kill wifi on startup when it's left in an inconsistent state by your launcher or something.

**Comment by NiLuJe** [comment]
As far as wifi strength is concerned, that's a peculiarity of Kobos in general.

**Comment by Biep** [comment]
Ah.  So I guess I'll be stuck with the apparent no-signal of the wifi.  For some (probably unrelated) reason I cannot connect to wifi right now (nickel refuses too), so I can't update the direct way.  Download the nightly and export the zip to .adds?

**Comment by NiLuJe** [comment]
It's exacerbated in *some* circumstances, generally having to do with having done stuff behind nickel's back.
It's more pronounced on some devices than others.

(i.e., on my H2O, after restarting Nickel, there's a high probability the signal icon will be entirely grayed-out. Much less so on the Forma).

**Comment by NiLuJe** [comment]
As far as manual updates go, [yep](https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices#what-about-manual-updates).

**Comment by Biep** [comment]
> 
> 
> Try the latest nightly it, it should attempt to kill wifi on startup when it's left in an inconsistent state by your launcher or something.

Yes!  The problem is still there, but restarting KOReader (directly from the sandwich menu) unchecked the "Wi-Fi connection" box.  I suppose stronger versions of restarting (through KSM, by switching off and on) will do the same,
Thank you!

**Comment by NiLuJe** [comment]
I'm confused. Does it (i.e., "Wi-Fi connection" behing checked without any actual working connection) still happen with you usual workflow? (i.e., let's say, for of sake of testing: cold boot -> KSM -> KOReader).

**Comment by Biep** [comment]
No - but whenever I try and fail to connect to wifi I get in that state, and need to restart KOReader to get out of it again.
It would be great if the code that now runs upon starting would also run when trying to uncheck "Wi-Fi connection" - and if possible even immediately whenever a wifi connection attempt fails.

**Comment by NiLuJe** [comment]
Hmm, that's interesting, and something I hadn't thought could be an issue
before. Could you describe what you usually mean by "fail to connect to
wifi", in details, so I can try to see where a recovery mechanism might fit
best ;).

**Comment by Biep** [comment]
Normally it starts with Sandwich -> Check for OTA update.  This asks me whether I want to switch wifi on, and (another bug) doesn't do anything after that.  I have to check again, with the wifi already on, actually to get an answer.
I get shown a list of wifi sources, choose one, type the wrong password, and don't get connected.  At that point I am in the "failed state".
Maybe I am already in the failed state _before_ choosing one of the sources in the list - I didn't try that.

**Comment by Frenzie** [comment]
> (another bug) doesn't do anything after that.

Not quite a bug, although it's something I'd like to see changed. There's this alpha implementation I don't like too much (but it does the trick on Android):

https://github.com/koreader/koreader/blob/011370882f3d6da5c8f59b9c90c006182918cfa2/frontend/ui/network/manager.lua#L72-L89

**Comment by NiLuJe** [comment]
Not in front of the code right now, so mostly a note to self: unload wifi
modules on connection failure...

**Comment by KenMaltby** [comment]
Might want to provide a "Try again?" option before shutting down the WiFi.  "Try again" could just jump to the scan again, giving the user a chance to try another AP.

**Comment by KenMaltby** [comment]
Hmm... the "Edit" button is not working, for me.  I would add that the user could also reselect his original AP and try to enter the proper password.

**Comment by NiLuJe** [comment]
We're already doing a double-scan since... a few weeks.

A "Try Again" with the same misspelled password wouldn't help here ;).

**Comment by pazos** [comment]
@NiLuJe: what about ignoring the wpa_supplicant file used by nickel/ksm and using one just for KOReader?

**Comment by NiLuJe** [comment]
Meh, I really don't want to have to type my machine-generated passwords in two different places ;).

**Comment by pazos** [comment]
So nuke nickel and you'll need to type only once. :smile:

**Comment by Biep** [comment]
Or allow a way to import a file with wifi information over usb.

**Comment by Biep** [comment]
Hey - does that mean that point has been done?

**Comment by NiLuJe** [comment]
Waiting for a full build to cook to test, because I'm too lazy to apply today's changes manually.

**Comment by pazos** [comment]
Err, can this change be made just for kobos. AFAICT your changes are applied on complete_callback and nuke wireless association with AP if internet is not available. 

If not I will need to test that on Cervantes too..

**Comment by NiLuJe** [comment]
We do need to reset the internal flags everywhere, but, yeah, the turnOffWiFi should be limited to devices where turnOn loads kernel modules.

**Comment by NiLuJe** [comment]
... which includes Cervantes ;).

**Comment by Sopheus** [comment]
Hello, 

Regard "Networks without WPA/WPA2 encryption are not supported", is there a workaround? I'm trying to connect to company WiFi and we have our internal network without WPA/WPA2 encryption, yet Internet is available.

**Comment by Frenzie** [comment]
The workaround would be to connect outside of KOReader I'm afraid.
